Add default pod template support

eddycharly opened the pull request:
This PR adds support for a default pod template through the default `ConfigMap`.

This is useful when one wants to apply a pod template to all task and pipeline runs, for example specifying a node selector, a security policy, etc...

When a pod template is specified for a `PipelineRun` or `TaskRun` it takes precedence over the default pod template.
Therefore, the default pod template only applies to `PipelineRun` or `TaskRun` that don't have a pod template.

# Submitter Checklist

- [x] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [x] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [ ] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

vdemeester commented:
> Yet to be defined : what behavior should we implement when a pod template and a default pod template are defined ? Merge both ? Pod template overrides completely the default one ? How can we bypass the default pod template if wanted (explicitly specifying `null` clears out the default pod template) ?

We should stick to what we do for the rest, aka, a Task step containerspec takes over the same field the stepTemplate field. In that case, a user using `podTemplate` in his TaskRun would take over the field from the "default" config `podTemplate`. And as for `stepTemplate` we would merge the fields.

eddycharly commented:
@vdemeester thanks for the feedback. i agree that sticking to what is already done is good for consistency.

Trying to build right now, my changes introduce a cyclic dependency config needs v1alpha1 (for `PodTemplate` struct) and v1alpha1 needs config (in pipelinerun_defaults.go for example).
How would you recommend to solve the issue ?

Should i create an interface ? A package to hold structs like `PodTemplate` ?

vdemeester commented:
> Trying to build right now, my changes introduce a cyclic dependency as config needs v1alpha1 (for `PodTemplate` struct), and v1alpha1 needs config (in pipelinerun_defaults.go for example).
> 
> How would you recommend solving this dependency issue ?
> 
> Should i create an interface ? A package to hold structs like `PodTemplate` ?

I think we have a little bit time to think about that. I would want to wait for other feedback on this feature before tackling this issue.

That said, on top of my head, we can extract the sturct into a package and use type aliasing to not have the cyclic dependency. I might take this approach for #1526 (more or less).

eddycharly commented:
Just trying to figure out what would have to be done to implement it... in case it is accepted ;)

Still waiting for feedback/discussion of course.

sbwsg commented:
> what behavior should we implement when a pod template and a default pod template are defined ? Merge both ?

I would suggest _not_ trying to perform a merge.  My preference would be that the pod gets either the one they specify or the default if not specified, not an amalgamation of both.

eddycharly commented:
@sbwsg @vdemeester should i go ahead with this or do you think it's useless ?

vdemeester commented:
@eddycharly I think it make sense :angel: and I think, as a start I do like @sbwsg suggestion. If there is no PodTemplate on a TaskRun, let's use the one from the default (if any), and if there is a PodTemplate, it overrides them all :angel:

vdemeester commented:
/ok-to-test

eddycharly commented:
@vdemeester to resolve the cyclic reference issue i had with pod.go, i moved it in `pkg/apis/pipeline` instead of `pkg/apis/pipeline/v1alpha1`.
Do you have a recommandation for the package where it should be or `pipeline` is fine ?

vdemeester commented:
> @vdemeester to resolve the cyclic reference issue i had with pod.go, i moved it in `pkg/apis/pipeline` instead of `pkg/apis/pipeline/v1alpha1`.
> Do you have a recommandation for the package where it should be or `pipeline` is fine ?

hum :thinking: why is there a cyclic reference issue ?

eddycharly commented:
It introduces a dependency from `config` to `v1alpha1` (to deserialize the pod template struct from the config map), but `v1alpha1` already depends on `config`...

eddycharly commented:
@vdemeester how can i generate the deepcopy functions ? `hack/update-codegens.sh` doesn't seem to regenerate the zz_generated.deepcopy.go files.

About the type alias, it seems to be working, thanks.

eddycharly commented:
@vdemeester quick question.
Not sure if i need to change this func in `default.go` to account for `DefaultPodTemplate` ?
```go
// Equals returns true if two Configs are identical
func (cfg *Defaults) Equals(other *Defaults) bool {
	return other.DefaultTimeoutMinutes == cfg.DefaultTimeoutMinutes &&
		other.DefaultServiceAccount == cfg.DefaultServiceAccount
}
```

I don't see it used anywhere.

vdemeester commented:
> @vdemeester quick question.
> Not sure if i need to change this func in `default.go` to account for `DefaultPodTemplate` ?
> 
> ```go
> // Equals returns true if two Configs are identical
> func (cfg *Defaults) Equals(other *Defaults) bool {
> 	return other.DefaultTimeoutMinutes == cfg.DefaultTimeoutMinutes &&
> 		other.DefaultServiceAccount == cfg.DefaultServiceAccount
> }
> ```
> 
> I don't see it used anywhere.

It is used "magically" by the webhook and `knative/pkg`, so yeah :angel:

eddycharly commented:
> It is used "magically" by the webhook and knative/pkg, so yeah 👼

Ok, i will update it to account for pod template then.

One more question please.

In `PipelineRunSpec` and `TaskRunSpec` struct i changed the `PodTemplate` property to be a pointer. I did that to easily check if a pod template was specified.

```go
	PodTemplate *PodTemplate `json:"podTemplate,omitempty"`
```

The concern i have is that client code probably doesn't check for nil before accessing the `PodTemplate` property.

For example, i think this code in the `test/builder` package is going to fail 
https://github.com/tektoncd/pipeline/blob/master/test/builder/task.go#L398.

Should i remove the pointer and check if the struct is _empty_ instead of checking for `nil` ? Or do you prefer the pointer version ?

PS: do you have a slack so that i can ask questions more easily ? ;)

sbwsg commented:
> In PipelineRunSpec and TaskRunSpec struct i changed the PodTemplate property to be a pointer. I did that to easily check if a pod template was specified.
> 
>     PodTemplate *PodTemplate `json:"podTemplate,omitempty"`
> The concern i have is that client code probably doesn't check for nil before accessing the PodTemplate property.

Tha's a good point!  Yeah this might be considered a breaking change?  What do you think @vdemeester?  It definitely makes things trickier if it's not a nil-able type :/

> do you have a slack so that i can ask questions more easily ? ;)

Yes absolutely!  Instructions here: https://github.com/tektoncd/community/blob/master/contact.md#slack

vdemeester commented:
> > The concern i have is that client code probably doesn't check for nil before accessing the PodTemplate property.
> 
> Tha's a good point! Yeah this might be considered a breaking change? What do you think @vdemeester? It definitely makes things trickier if it's not a nil-able type :/

So we have to distinguish between API breaking change and go breaking change. We definitely want to reduce/remove any API breaking change (by doing things backward compatible), but go breaking change might happen more often. This change doesn't break the API, it will just "force" the user of tektoncd/pipeline as a library to update some of their code, but it shouldn't prevent previous version of the code to still be able to marshall/unmarshall the object ; so I am fine with it :angel:.

eddycharly commented:
Ok, fine 👍 
I fixed the `test/builder` package and added tests.

I'm waiting for https://github.com/tektoncd/pipeline/pull/1853 to be merged, then i will add a bit of doc about the default pod template feature.
I still have to add a few more tests and it should be done.

eddycharly commented:
/test pull-tekton-pipeline-build-tests

eddycharly commented:
/assign @bobcatfish

vdemeester commented on config/config-defaults.yaml:
```diff
@@ -41,3 +41,8 @@ data:
+    # default-pod-template contains the default pod template to use
+    # TaskRun and PipelineRun, if none is specified. If a pod template
+    # is specified, the default pod template is ignored.
+    default-pod-template:
```
nit: Should it be commented out by default ? instead of empty ? (I don't think it has any impact :sweat_smile: but…)

eddycharly replied:
done

vdemeester commented on docs/install.md:
```diff
@@ -181,24 +180,33 @@ creation of a persistent volume could be slower than uploading/downloading files
 the persistent volume can fail.
 
-### Overriding  default ServiceAccount used for TaskRun and PipelineRun
+### Overriding default ServiceAccount, Timeout or PodTemplate used for TaskRun and PipelineRun
```
:+1:

vdemeester commented on hack/update-codegen.sh:
```diff
@@ -46,6 +46,11 @@ ${GOPATH}/bin/deepcopy-gen \
   --go-header-file ${REPO_ROOT_DIR}/hack/boilerplate/boilerplate.go.txt \
   -i github.com/tektoncd/pipeline/pkg/apis/config
 
+${GOPATH}/bin/deepcopy-gen \
```
ah this is needed :sweat_smile:

vdemeester commented on pkg/apis/config/default.go:
```diff
@@ -21,6 +21,10 @@ import (
 	"strconv"
 	"time"
 
+	"gopkg.in/yaml.v2"
```
can we group those imports ?

eddycharly replied:
done

eddycharly replied:
done

eddycharly commented on pkg/apis/pipeline/pod/template.go:
```diff
@@ -0,0 +1,105 @@
+	// Selector which must match a node's labels for the pod to be scheduled on that node.
+	// More info: https://kubernetes.io/docs/concepts/configuration/assign-pod-node/
+	// +optional
+	NodeSelector map[string]string `yaml:"nodeSelector,omitempty" json:"nodeSelector,omitempty"`
```
@vdemeester i had to add yaml tags on the `Template` struct to be able to deserialize it. But the yaml tags were not present on the other structs, is this normal ? Am i using the wrong serialization tool ?

vdemeester replied:
:thinking: They shouldn't be needed indeed…

eddycharly replied:
if i don't put them `NodeSelector` is expected to be a property named `nodeselector` by `yaml.Unmarshal` rather than `nodeSelector`

vdemeester replied:
:thinking: We should use kubernetes's unmarshal mechanism then (I think they do expose it from somewhere)

eddycharly replied:
i thought it was something like this, i'll change that

eddycharly replied:
Looks like you're using `ghodss/yaml` to unmarshal.
I changed the code to use it and removed the yaml tags on the `Template` struct.

The `ghodss/yaml` you use is in vendor and does not support UnmarshalStrict, i commented out some tests that checked unmarshaling errors. (these are new tests i added)

eddycharly commented:
Outch.. conficts :(
Can i resolve conflicts and someone merge the PR please ? @vdemeester @bobcatfish

bobcatfish commented:
Thanks for keeping at this @eddycharly !

To resolve the conflicts you'll need to rebase your branch onto the latest version of master and then force push, then we can lgtm and merge. Let us know if you want us to point you in the right direction to do that!

eddycharly commented:
/test pull-tekton-pipeline-integration-tests

eddycharly commented:
@bobcatfish i resolved the conflicts.

bobcatfish commented:
```
Something went wrong: failed to prepare test environment: --provider=gke boskos failed to acquire project: resources not found
```

This is happening over in https://github.com/tektoncd/pipeline/pull/1888 as well, re-opened https://github.com/tektoncd/plumbing/issues/29

eddycharly commented:
boskos seems to be out of resources :(

eddycharly commented:
/test pull-tekton-pipeline-integration-tests

bobcatfish commented:
Sorry about this @eddycharly trying to figure out what's going on :sob:

eddycharly commented:
Thanks for your support @bobcatfish !
Good luck, i will retry later, better let boskos alone for now ;)

eddycharly commented:
/test pull-tekton-pipeline-integration-tests

bobcatfish commented:
/test pull-tekton-pipeline-integration-tests

eddycharly commented:
/test pull-tekton-pipeline-integration-tests

eddycharly commented:
/test pull-tekton-pipeline-integration-tests

eddycharly commented:
@bobcatfish all tests are passing now. Can you give a look please ?

vdemeester commented:
@eddycharly can you squash your commits ?

eddycharly commented:
/test tekton-pipeline-unit-tests

eddycharly commented:
@vdemeester any idea why unit tests are failing ?
`I0120 11:06:38.619] 2020-01-20T11:06:37.193Z	ERROR	TestReconcile/success-with-cluster-task	taskrun/controller.go:59	Failed to create taskrun metrics recorder running_taskruns_count: cannot register view "running_taskruns_count"; a different view with the same name is already registered`

vdemeester commented:
@eddycharly interesting… I thought we fixed that test… There might be a slight race in there still… :thinking:
